api/src/main/java/org/openmrs/module/idgen/processor/SequentialIdentifierGeneratorProcessor.java

```diff
@@ -41,15 +43,21 @@ public void setIdentifierSourceService(IdentifierSourceService identifierSourceS
 	public synchronized List<String> getIdentifiers(IdentifierSource source, int batchSize) {
 		
 		SequentialIdentifierGenerator seq = (SequentialIdentifierGenerator) source;
+
+		// Get the next sequence value.
         Long sequenceValue = identifierSourceService.getSequenceValue(seq);
-    	if (sequenceValue == null || sequenceValue < 0) {
-    		if (seq.getFirstIdentifierBase() != null) {
-    			sequenceValue = IdgenUtil.convertFromBase(seq.getFirstIdentifierBase(), seq.getBaseCharacterSet().toCharArray());
-    		}
-    		else {
-    			sequenceValue = 1L;
-    		}
-    	}
+
+        // Get the first sequence value allowed based on the identifier base
+		Long baseSequenceValue = 1L;
+		String identifierBase = seq.evaluateFirstIdentifierBase();
+		if (StringUtils.isNotBlank(identifierBase)) {
+			baseSequenceValue = IdgenUtil.convertFromBase(identifierBase, seq.getBaseCharacterSet().toCharArray());
+		}
+
+		// If the base sequence value is greater than the next sequence value, reset the next sequence value to the base

```

Comment:
I pretty much trust you, but I'm not quite 100% getting how this works?

Reply:
Yeah, for this I don't want you to trust me.  This is too important to afford to screw up.

The "firstIdentifierBase" property of a SequentialIdentifierGenerator is there to essentially tell the generate what sequence number to start with.  For a simple incremental base10 generator, this would basically match what you want to initialize the "next sequence value" to, as a sequence of 1000 represents an identifier "1000".  

But with other bases, this is less straightforward.  For example, if you have a base16 generator, and want to start at A12, you'd configure the "firstIdentifierBase" as "A12", and the IdgenUtil.convertFromBase method computers that this identifier would be created with sequence number (2x1) + (1x16) + (10 x 256) = 2578, and it would ensure that if the sequence hadn't been used yet (eg. next sequence number is null), it would set it to this first number (2578) to ensure that the first identifier produced is "A12".

Up until now, this was only done if the existing firstIdentifierBase is null, as the assumption was that once a sequence is in use, the firstIdentifierBase is no longer necessary - we just increment from the last produced identifier.

What this change does is it allows the "firstIdentifierBase" to change based on evaluated criteria.  So rather than being a static value of "1000" to start the sequence at 1000, it has a "dynamic" value of, say, YYMM1000.  So as the current date changes, the base number can change.

So what the code here is doing is - rather than only checking a static first identifier base if the existing sequence is null, and setting the sequence to it - it evaluates the first identifier base every time, and if this is greater than what the next sequence would be, it moves the next sequence forward to match this evaluated first identifier base value.

That's the idea anyway...

Reply:
Ah, got it, that makes sense... it's a good way of handling the issue, but not sure I'm totally in love with it... :)

So, in the KGH case, our configuration would be:

prefix:KGH
base: ${currentDate:YYMM}0000

Correct?

Without thinking of the implementation details, if you asked me to how to define that identifier format from a "business" perspective, I would think it would be:

prefix:KGH${currentDate:YYMM}
base:0000

ie, I intuitively think of the date as being part of the prefix?  Or at least not part of the base?

The practical side of this is that this implementation wouldn't do much good if they came back and said they wanted the format to be:

KGH-YY-MM-XXXX

... which would seem to be a reasonable thing somebody might want...

Do we keep a history of the identifiers issues and in order?  One thing could be to support a prefix like:
KGH-${currentDate:YY}-${currentDate:MM}

... and then do a test where if previous prefix != current prefix, reset the sequential counter to the initial base?

Reply:
@mogoodrich thanks, good points. That was exactly how I originally tried to solve the problem- by making the date a component of the prefix- not sure if my original PR is still accessible, but that’s what I did. And when I realized that the solution addressed the date prefix but did not address the counter, I did consider the approach you are indicating- but didn’t end up going with it as it was unclear how to fit it into the existing data model.    Let’s brainstorm together a bit